### PR TITLE
feat(edge): add mTLS support [EE-2617]

### PR DESCRIPTION
**Pull request by cheloRydel**
Closes [EE-2617](https://portainer.atlassian.net/browse/EE-2617)

Replaces https://github.com/portainer/agent/pull/249 and https://github.com/portainer/agent/pull/234

Add support for mTLS configuration.
Goes with the corresponding change in Portainer, CE doesn't actually use mTLS but sets dependencies used in the EE PR, and has the same refactors/code improvements for codebase sync between both repos:
- CE: https://github.com/portainer/portainer/pull/6612
- EE: https://github.com/portainer/portainer-ee/pull/1286

Changes:
- Add `SSLCert`, `SSLKey` and `SSLCACert` options, which allow a path to be passed for bind-mounted certificates to be used. When provided, the certificates are configured in the transport level of every outgoing http request.
- Refactor `NewPortainerClient` function to receive `httpClient` as parameter, also remove `edgeInsecurePoll`attribute from `PollService` to simplify code.
- Add `dev-scripts/mtlscerts.sh` with sample certificate scripts, meant for testing.
- Rename `EdgeServerAddr`and `EdgeServerPort` to `EdgeUIServerAddr`and `EdgeUIServerPort`, for clarity.

How to run locally:
To use this, I create the 3 certificates we need (they go into ~/.config/portainer/certs/) using the `./dev-scripts/mtlscerts.sh` script, and then build and start both portainer and the agent using those.

I then start portainer with the following args:

```
       --sslcert
       ${HOME}/.config/portainer/certs/server-cert.pem
       --sslkey
       ${HOME}/.config/portainer/certs/server-key.pem
       --sslcacert
       ${HOME}/.config/portainer/certs/ca.pem
       --data
       ${HOME}/portainer-data
       --assets
       ${HOME}/code/portainer/portainer/dist
```

And an agent with (using sample `EdgeKey` and `EdgeID` here):

```
        --edge
        --edge-key
        aHR0cHM6Ly8xMjcuMC4wLjE6OTQ0M3wxMjcuMC4wLjE6ODAwMHxlZDo4YzpiNDozNjozZTpiYzo0MzozMTo4MTpkYzplMDoyZDozYzo4ZDoyZDoxZXw1NQ
        --edge-id
        7e2b0143-c511-43c3-844c-a7a91ab0bedc
        --sslcert
        ${HOME}/.config/portainer/certs/agent-cert.pem
        --sslkey
        ${HOME}/.config/portainer/certs/agent-key.pem
        --sslcacert
        ${HOME}/.config/portainer/certs/ca.pem
        --assets
        ./
        --data
        ${HOME}/portainer-data-agent
```

### CODE REVIEW

File: dev-scripts/mtlscerts.sh [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#!/bin/bash
+
+# very much just copied from https://lemariva.com/blog/2019/12/portainer-managing-docker-engine-remotely
+# production use should involve a real external certificate management system
+
+export CURRENT=$(pwd)
+export HOST=portainer.p1.alho.st
+export CERTDIR=~/.config/portainer/certs/
```
Comment (chiptus) [inline dev-scripts/mtlscerts.sh:8]: maybe recieve that as a parameter? 
```suggestion
export CERTDIR=${1:-"~/.config/portainer/certs/"}
```

**Review by chiptus — APPROVED** [review]
LGTM

btw, what about adding to the main `./dev.sh` something like `./dev.sh ssl` to generate the certs?

**Comment by cheloRydel** [comment]
> 

I'd say to keep those separate, while testing I wanted to be able to generate the certs without having to start the portainer server as well

**Comment by chiptus** [comment]
> > 
> 
> I'd say to keep those separate, while testing I wanted to be able to generate the certs without having to start the portainer server as well

./dev.sh doesn't start anything, it's just help utiltiles (deploy/compile)...

**Comment by cheloRydel** [comment]
> > > 
> > 
> > 
> > I'd say to keep those separate, while testing I wanted to be able to generate the certs without having to start the portainer server as well
> 
> ./dev.sh doesn't start anything, it's just help utiltiles (deploy/compile)...

my bad, I was look at the portainer script, not the agent

File: http/server.go [hunk 2]
```diff
@@ -136,7 +136,7 @@ func (server *APIServer) StartSecured(edgeMode bool) error {
 	log.Printf("[INFO] [http] [server_addr: %s] [server_port: %s] [secured: %t] [api_version: %s] [message: Starting Agent API server]", server.addr, server.port, config.Secured, agent.Version)
 
 	tlsConfig := &tls.Config{
-		MinVersion: tls.VersionTLS12,
+		MinVersion: tls.VersionTLS13,
 		CipherSuites: []uint16{
```
Comment (andres-portainer) [inline http/server.go]: The CipherSuites should be removed for TLS 1.3, they are not configurable.

File: dev-scripts/mtlscerts.sh [hunk 3]
```diff
@@ -0,0 +1,52 @@
+#!/bin/bash
+
+# very much just copied from https://lemariva.com/blog/2019/12/portainer-managing-docker-engine-remotely
+# production use should involve a real external certificate management system
+
+export CURRENT=$(pwd)
+export HOST=portainer.p1.alho.st
```
Comment (andres-portainer) [inline dev-scripts/mtlscerts.sh]: Maybe take this as an argument for the script?

**Review by andres-portainer — APPROVED** [review]

File: dev.sh [hunk 4]
```diff
@@ -53,6 +55,8 @@ case $1 in
     podman)
         deploy_command -p -c "${@:2}"
     ;;
+    mtlscerts)
```
Comment (chiptus) [inline dev.sh]: since mtlscerts expose a script and not a function, this will run it anyway (on any run of `./dev.sh`).

I'd say, remove this from dev.sh

**Review by chiptus — CHANGES_REQUESTED** [review]
remove mtls from dev script

File: dev-scripts/mtlscerts.sh [hunk 5]
```diff
@@ -0,0 +1,54 @@
+#!/bin/bash
+
+# very much just copied from https://lemariva.com/blog/2019/12/portainer-managing-docker-engine-remotely
+# production use should involve a real external certificate management system
+
+function mtlscerts_command() {
+  export CURRENT=$(pwd)
+  export HOST=${1:-"portainer.p1.alho.st"}
+  export CERTDIR=~/.config/portainer/certs/
```
Comment (chiptus) [inline dev-scripts/mtlscerts.sh]: these should be `local`, as you don't need them outside of this script
```suggestion
  local CURRENT=$(pwd)
  local HOST=${1:-"portainer.p1.alho.st"}
  local CERTDIR=~/.config/portainer/certs/
```

and also convention says to name local variables with lower caps, but it's not as important

**Review by chiptus — APPROVED** [review]

**Review by deviantony — APPROVED** [review]
LGTM

**Pull request merged**